lib/composite_cache_store.rb

```
@@ -4,146 +4,157 @@
+  attr_accessor :logger
 
   # Returns a new CompositeCacheStore instance
-  def initialize(*layers)
-    if layers.blank?
-      layers << ActiveSupport::Cache::MemoryStore.new(DEFAULT_LAYER_1_OPTIONS)
-      layers << ActiveSupport::Cache::MemoryStore.new(DEFAULT_LAYER_2_OPTIONS)
-    end
+  def initialize(options = {})
+    options = options.dup || {}
+    layers = options.delete(:layers) || []
 
-    message = "All layers must be instances of ActiveSupport::Cache::Store"
-    layers.each do |layer|
-      raise ArgumentError.new(message) unless layer.is_a?(ActiveSupport::Cache::Store)
-    end
+    raise ArgumentError.new("A layered cache requires more than 1 layer!") unless layers.size > 1
 
-    layers.freeze
-    @layers = layers
+    stores = layers.select { |layer| layer.is_a? ActiveSupport::Cache::Store }
+    raise ArgumentError.new("All layers must be instances of ActiveSupport::Cache::Store!") unless stores.size == layers.size
+
+    @layers = layers.freeze
+    @logger = options[:logger]
+    @options = options
   end
 
   def cleanup(...)
-    layers.each { |store| store.cleanup(...) }
+    layers.each { |layer| layer.cleanup(...) }
   end
 
   def clear(...)
-    layers.each { |store| store.clear(...) }
+    layers.each { |layer| layer.clear(...) }
+  end
+
+  def increment(name, amount = 1, options = nil)
+    provisional_layers.each { |layer| layer.delete(name, options) }
+    layers.last.increment(name, amount, options)
   end
 
-  def decrement(...)
-    layers.each { |store| store.decrement(...) }
+  def decrement(name, amount = 1, options = nil)
+    provisional_layers.each { |layer| layer.delete(name, options) }
+    layers.last.decrement(name, amount, options)
   end
 
   def delete(...)
-    layers.each { |store| store.delete(...) }
+    layers.each { |layer| layer.delete(...) }
   end
 
   def delete_matched(...)
-    layers.each { |store| store.delete_matched(...) }
+    layers.each { |layer| layer.delete_matched(...) }
```

Many caches won't support this rarely-invoked method, and I suppose `CompositeCacheStore` could rescue any `NotImplementedError`s they might raise to return one of its own, but it's probably fine to just let the first layer raise. 👍

---

Agree we can just let this bubble up.